### PR TITLE
Removing bootstrap: admin sidebar navigation

### PR DESCRIPTION
**Pull request by ludwiczakpawel**
## What type of PR is this? (check all applicable)

- [x] Refactor

## Description

This PR is another step towards removing Bootstrap. This one focuses on sidebar navigation: it basically removes the JS-based collapsing/expanding behavior on 2nd level navigation which btw used to be pretty glitchy and the use of JS was simply pointless there since we always reload the page anyway.

## QA Instructions, Screenshots, Recordings

Go to /admin and make sure navigation works as expected.

### UI accessibility concerns?

There's still concern about indicating 2nd level navigation for screen reader users. FWIW it's not worse than what we had :D but I believe @aitchiss is going to give it a look. 

## Added/updated tests?

I did remove some tests because they were made for JS-behavior which is now gone.

## [Forem core team only] How will this change be communicated?

It won't.

### CODE REVIEW

File: app/lib/menu.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ def scope(name, svg, children)
     @items[name] = { svg: "#{svg}.svg", children: children }
   end
 
-  def item(name:, controller: name, children: [], visible: true)
```
Comment (ludwiczakpawel) [inline app/lib/menu.rb:20]: @Ridhwana I made some changes here and would love to know if this is a right approach.

Reason for these changes is I wanted to always have a clear indication re what nav item is the "parent" one so we can highlight the right item without hacking too much in the view.

File: app/views/admin/shared/_nested_sidebar.html.erb [hunk 2]
```diff
@@ -1,37 +1,35 @@
+<li>
+  <a class="c-link c-link--block c-link--icon-left" href="<%= admin_path %>" aria-current="<%= "page" if controller.controller_name == "overview" %>">
+    <%= crayons_icon_tag("stack-line", class: "c-link__icon") %>
+    Overview
+  </a>
+</li>
```
Comment (ludwiczakpawel) [inline app/views/admin/shared/_nested_sidebar.html.erb:6]: @Ridhwana it feels naughty to me that this link is the special snowflake being outside admin_menu.rb but I'm not sure how to add it there.. Thoughts?

File: app/views/admin/shared/_nested_sidebar.html.erb [hunk 3]
```diff
@@ -1,55 +1,45 @@
...
+        <%= crayons_icon_tag("stack-line", class: "c-link__icon") %>
+        Overview
+      </a>
+    </li>
+
+    <% menu_items.each do |group_name, group| %>
+      <% current = deduced_scope(request).to_s == group[:children][0][:parent].to_s %>
+      <% children = group[:children].length == 1 %>
+      <% path = children ? group[:children][0][:controller] : "#{group_name}/#{group[:children].detect { |child| child[:visible] }[:controller]}" %>
+      <% path = "#{admin_path}/#{path}" %>
```
Comment (ludwiczakpawel) [inline app/views/admin/shared/_nested_sidebar.html.erb:14]: @Ridhwana i did this to make things more readable in the view as it was previously all over the place... But I feel this is not the most appropriate place or this?

File: app/views/admin/shared/_nested_sidebar.html.erb [hunk 4]
```diff
@@ -1,55 +1,45 @@
...
-           data-action="click->sidebar#expandDropdown">
-          <%= crayons_icon_tag(group[:svg], class: "dropdown-icon", aria_hidden: true) %>
+<nav>
+  <ul class="m-0">
+    <li>
+      <a class="c-link c-link--block c-link--icon-left" href="<%= admin_path %>" aria-current="<%= "page" if controller.controller_name == "overview" %>">
+        <%= crayons_icon_tag("stack-line", class: "c-link__icon") %>
+        Overview
+      </a>
+    </li>
```
Comment (ludwiczakpawel) [inline app/views/admin/shared/_nested_sidebar.html.erb:8]: @Ridhwana it feels naughty to me that this link is the special snowflake being outside admin_menu.rb but I'm not sure how to add it there.. Thoughts?

File: app/views/admin/shared/_nested_sidebar.html.erb (hunk 3 above)
Comment (Ridhwana) [reply app/views/admin/shared/_nested_sidebar.html.erb:14]: @ludwiczakpawel yeah you can probably add it as helper methods `/admin/helpers`

File: app/lib/menu.rb (hunk 1 above)
Comment (Ridhwana) [reply app/lib/menu.rb:20]: Thanks Pawel, here's my thoughts:

When we define the structure, we already have a hierarchy and a "children" key. This shows who the parent is based on the data structure:
```
{:people=>{:svg=>"group-2-line.svg", :children=>[{:name=>"people", :controller=>"users", :children=>[], :visible=>true}]},
 :content_manager=>
  {:svg=>"dashboard-line.svg",
   :children=>
    [{:name=>"posts", :controller=>"articles", :children=>[], :visible=>true},
     {:name=>"comments", :controller=>"comments", :children=>[], :visible=>true},
     {:name=>"badges",
      :controller=>"badges",
      :children=>[{:name=>"library", :controller=>"badges", :children=>[], :visible=>true}, {:name=>"achievements", :controller=>"badge_achievements", :children=>[], :visible=>true}],
      :visible=>true},
     {:name=>"organizations", :controller=>"organizations", :children=>[], :visible=>true},
     {:name=>"podcasts", :controller=>"podcasts", :children=>[], :visible=>true},
     {:name=>"tags", :controller=>"tags", :children=>[], :visible=>true}]},
 :customization=>
  {:svg=>"tools-line.svg",
   :children=>
    [{:name=>"config", :controller=>"config", :children=>[], :visible=>true},
     {:name=>"html variants", :controller=>"html_variants", :children=>[], :visible=>true},
     {:name=>"display ads", :controller=>"display_ads", :children=>[], :visible=>true},
     {:name=>"navigation links", :controller=>"navigation_links", :children=>[], :visible=>true},
     {:name=>"pages", :controller=>"pages", :children=>[], :visible=>true},
     {:name=>"profile fields", :controller=>"profile_fields", :children=>[], :visible=>false}]},
 :admin_team=>{:svg=>"user-line.svg", :children=>[{:name=>"admin team", :controller=>"permissions", :children=>[], :visible=>true}]},
 :moderation=>
  {:svg=>"mod.svg",
   :children=>
    [{:name=>"reports", :controller=>"reports", :children=>[], :visible=>true},
     {:name=>"mods", :controller=>"mods", :children=>[], :visible=>true},
     {:name=>"moderator actions ads", :controller=>"moderator_actions", :children=>[], :visible=>true},
     {:name=>"privileged reactions", :controller=>"privileged_reactions", :children=>[], :visible=>true}]},
 :advanced=>
  {:svg=>"flashlight-line.svg",
   :children=>
    [{:name=>"broadcasts", :controller=>"broadcasts", :children=>[], :visible=>true},
     {:name=>"response templates", :controller=>"response_templates", :children=>[], :visible=>true},
     {:name=>"sponsorships", :controller=>"sponsorships", :children=>[], :visible=>true},
     {:name=>"developer tools",
      :controller=>"tools",
      :children=>
       [{:name=>"tools", :controller=>"tools", :children=>[], :visible=>true},
        {:name=>"vault secrets", :controller=>"secrets", :children=>[], :visible=>true},
        {:name=>"data update scripts", :controller=>"data_update_scripts", :children=>[], :visible=>false}],
      :visible=>true}]},
 :apps=>
  {:svg=>"palette-line.svg",
   :children=>
    [{:name=>"consumer apps", :controller=>"consumer_apps", :children=>[], :visible=>true},
     {:name=>"listings", :controller=>"listings", :children=>[], :visible=>true},
     {:name=>"welcome", :controller=>"welcome", :children=>[], :visible=>true}]}}
```
Hence, to me this feels redundant and an overhead where we need to keep track of the parent when it's clear in the data structure.

File: app/views/admin/shared/_nested_sidebar.html.erb [hunk 5]
```diff
@@ -1,55 +1,45 @@
...
+    <% menu_items.each do |group_name, group| %>
+      <% current = deduced_scope(request).to_s == group[:children][0][:parent].to_s %>
+      <% children = group[:children].length == 1 %>
+      <% path = children ? group[:children][0][:controller] : "#{group_name}/#{group[:children].detect { |child| child[:visible] }[:controller]}" %>
+      <% path = "#{admin_path}/#{path}" %>
+      <li>
+        <a
+          class="c-link c-link--block c-link--icon-left <%= "c-link--current" if current %>"
+          aria-current="<%= "page" if current && children %>"
+          href="<%= path %>">
```
Comment (nickytonline) [inline app/views/admin/shared/_nested_sidebar.html.erb:19]: Does the `c-link--current` CSS class indicate this is the active section and if it happens to have child elements, then it's the current page? Just trying to make sure I understand the logic as `[aria-current="page"]` is typically enough to style the link in navigation if that link is for the current page.

Comment (ludwiczakpawel) [reply app/views/admin/shared/_nested_sidebar.html.erb:19]: That is correct, sir.

Comment (ludwiczakpawel) [reply app/views/admin/shared/_nested_sidebar.html.erb:19]: if there's no children, i use aria-current. if there are children, i use css class to visually indicate active prent and then use aria-current for appropriate 2nd level nav link.

File: app/lib/menu.rb (hunk 1 above)
Comment (ludwiczakpawel) [reply app/lib/menu.rb:20]: 👍 I removed redundant data.

File: app/helpers/admin_helper.rb [hunk 6]
```diff
@@ -14,4 +14,21 @@ def display_name(group_name)
   def dom_safe_name(group_name)
     group_name.gsub(/\s+|\./, "_").gsub(/\A(\d)/, '_\1')
   end
+
+  def current?(request, group, group_name)
+    comparable = group[:children][0][:parent] ? group[:children][0][:parent].to_s : group_name.to_s
```
Comment (citizen428) [inline app/helpers/admin_helper.rb:19]: Unless I'm misreading your intention, this should do the same:

```suggestion
    comparable = (group[:children][0][:parent] || group_name).to_s
```

**Review by citizen428 — APPROVED** [review]

File: app/helpers/admin_helper.rb [hunk 7]
```diff
@@ -14,4 +14,20 @@ def display_name(group_name)
   def dom_safe_name(group_name)
     group_name.gsub(/\s+|\./, "_").gsub(/\A(\d)/, '_\1')
   end
+
+  def current?(request, group, group_name)
+    deduced_scope(request).to_s == (group[:children][0][:parent] || group_name).to_s
+  end
+
+  def children?(group)
+    group[:children].length == 1
```
Comment (citizen428) [inline app/helpers/admin_helper.rb:23]: Just double-checking as I'm not 100% sure what the data structure looks like, but we really want `== 1` and not `>= 1`, right?

Comment (ludwiczakpawel) [reply app/helpers/admin_helper.rb:23]: I think this helper name was a bit misleading here and its usage in `_nested_sidebar.html.erb` was off too.. I changed that logic a little so I think it makes more sense now :). But that's a good catch, thanks!

File: app/views/admin/shared/_nested_sidebar.html.erb (hunk 4 above)
Comment (Ridhwana) [reply app/views/admin/shared/_nested_sidebar.html.erb:8]: @ludwiczakpawel i have this halfway working (except for showing it as the current link when we land on the page), I'll fix it in a subsequent PR.

**Review by Ridhwana — APPROVED** [review]
I think it may have been useful to add some cypress testing to test the behaviour the sidebar here (i.e. ensure that the correct menu item section is opening when we hit a page etc.). Maybe it can be added to the small wins board? 

I'll also add a small task for myself to get the overview link into the admin_menu item.

**Review by aitchiss — APPROVED** [review]
All works as expected 👍 

There's definitely an accessibility issue here... I think these changes do make it a little more confusing for screen reader users as we've lost the `aria-expanded` clue which gave a hint that something might now be expanded on the new page. But I think the whole thing needs thought about separately when we have time to do so properly.

The problem is that we are using a link in some cases to trigger an expanding menu, and also direct to a _different_ page than the link text indicates (e.g. there is no "content manager" page, it takes you to "posts"). 

For a screen reader user this means there's no indication when a sub menu has expanded. It also means some links appear to be broken (e.g. I follow the 'content manager' link and end up on 'posts' - without the visual cue of seeing a menu now expanded, you're likely to just think the link took you to the wrong place). For all users, it means behaviour is inconsistent (there's no visual clue which have sub-menus and which don't).

I don't really see a way around this a11y problem without rethinking the design so that a sub-menu can be expanded _while staying on the same page_. But that's a bigger job and not in the scope of just removing bootstrap 😄

**Pull request merged**